README.txt
$Id: README.txt,v 1.3 2001/08/28 06:46:36 boisvert Exp $

Package readme for: jdbm.*
Package version: see VERSION
Package status: beta
Package URL: http://jdbm.sourceforge.net
Package License: BSD


JDBM Project


WHAT IT IS
==========

- Easy API to persist Java objects (or byte arrays).  
- Basic layers are meant to ressemble the popular GDBM API.
- Transaction support for fail-safe operation.



WHAT IT'S *NOT*
===============

- NOT a multiuser database
- NOT compatible with existing GDBM databases
- NOT an SQL database (no query capabilities)
- NOT a full-blown object-oriented database
  (missing query, locking, schema, garbage collecting, ...)


FEATURES
========

- Simple Hashtable-like interface.

- 100% Java code.

- Fast: Uses no persistent rollback buffer.  An in-memory transaction log is
  kept.  (It can be disabled for even greater performance.)

- Transactional:  Updates are garanteed to be atomic and durable once commited.

- Auto-recovery.  If your application crashes, the next time your application
  opens the database, it will automatically recover any completed transactions.

- Allows implicit or explicit transaction demarcation (commit, rollback)
  through a low-level API called "jdbm.recman.RecordManager".

- (Planned) Support for JTA (Java Transaction API) to allow participation 
  in distributed transactions.


LIMITATIONS
===========

- Size of objects and transaction is limited to available memory size.

- Practical limit on database size is set by your operating system.


CONTRIBUTORS
============

Cees de Groot <cg@cdegroot.com> wrote the original RecordManager 
implementation and started the project.

Alex Boisvert <boisvert@intalio.com> added GDBM-like interface, 
HTree and BTree implementations, caching, rollback functionality 
and some bug fixing.


CHANGES.txt
$Id: CHANGES.txt,v 1.11 2002/10/13 18:54:42 boisvert Exp $

CHANGE LOG
==========

For version "CVS":

    * Better abstraction for RecordManager, which is now an interface.
      A concrete RecordManager instance is now obtained from
      RecordManagerFactory (and RecordManagerFactory supports 'pluggable'
      RecordManager providers)

    * Faster log checkpointing by avoiding unnecessary writes to same block
      in TransactionManager (contributed by Nathan Clement)

    * Added ability to set split point in BTree to optimize insert performance
      of entries with monotonically increasing/decreasing keys.  Also added
      statistics on BTree to monitor performance.
      (contributed by Kent Fitch - kentfitch@hotmail.com)

    * Renamed methods called 'assert' which is a reserved word in JDK 1.4+.
      (suggested by Yonas Jongkind)

    * Performance improvement in BlockIo.readLong() contributed by
      Erwin Bolwidt <ejb@klomp.org>

    * Switched to using java.util.Comparator interface instead of
      JDBM-specific comparator.
     

For version 0.13:

    * Upgraded to Ant 1.4.1 (and switched to Crimson instead of Xerces
      XML Parser).  Updated the build file accordingly and fixed a number
      of build issues.

    * Upgraded to JUnit 3.7.  Corrected use of deprecated methods in test
      harness.

    * Added IntegerComparator and updated TestBTree unit test
      (both contributed by Christof Dallermassl)

    * Fixed TestBTree.testBasics() and TestBTree.multiThreadAccess()
      test cases.

    * RecordManager now throws IOException if used after being closed()

    * Updated BTree.browse() javadoc to warn against structural modifications
      to the BTree while browsing.


For version 0.12:
    * Fixed a corruption bug in B+Tree during insertion.
    * Added {JDBMHashtable, BTree, HTree}.getRecid() methods
    * Loading of an existing HTree now uses static HTree.load() method
      instead of constructor
    * Added synchronization to RecordManager methods for thread-safeness.


For version 0.11:
    * Added new B+Tree data structure implementation.

For version 0.10:
    * Bug fix release with performance enhancements.
    * Fixed an infinite loop bug in Enumeration returned by HashDirectory
      (reported by Daniel Herlemont).  Added HashtableTest regression test.
    * Fixed LogicalRowIdManager.fetch() which dirtied blocks unnecessarily.
    * RecordFile.commit() optimized to avoid disk writes & sync() if no
      blocks have been dirtied in the transaction.

For version 0.09:
    * This is a major upgrade in terms of functionality.  We now have
      a GDBM-like API (actually, very much like Hashtable) and added
      explicit commit/rollback functionality.
    * Repackaging com.cdegroot.recman.* to jdbm.recman.*
    * Added jdbm.* package -- GDBM-like access
    * Added jdbm.hash.* package -- HTree implementation and corresponding
      test cases.
    * Changes in jdbm.recman.RecordManager to support commit() and rollback().
      This change affected many classes in jdbm.recman.*, notably
      TransactionManager.   Added new test case for rollback().
    * Test framework now integrated in the build.
    * Updated most of the code to reflect the new BSD license.

For version 0.081:
    * Important bug fix.  TransactionManager had a bug which could potentially
      corrupt the database file when the database was not closed explicitly.
      See bug #103848 on SourceForge.
      http://sourceforge.net/bugs/?func=detailbug&bug_id=103848&group_id=4155
      Bug report & fix courtesy of Alex Boisvert [boisvert@exoffice.com].

For version 0.08:
    * Cleanups. Cleaning up recman/FileHeader seems to have affected
      performance, but not bad enough for now (we can always optimize
      it again).
    * Added support for "root" rowids, which are special rowids that
      are accessed by an index and stored in the file header. Can be
      used for bootstrapping access to other structures.

For version 0.07:
    * Updated license to GNU Library General Public License.
    * More verbosity in doc comments on recman/RecordManager.



src/main/jdbm/RecordManager.java
/**
 * JDBM LICENSE v1.00
 *
 * Redistribution and use of this software and associated documentation
 * ("Software"), with or without modification, are permitted provided
 * that the following conditions are met:
 *
 * 1. Redistributions of source code must retain copyright
 *    statements and notices.  Redistributions must also contain a
 *    copy of this document.
 *
 * 2. Redistributions in binary form must reproduce the
 *    above copyright notice, this list of conditions and the
 *    following disclaimer in the documentation and/or other
 *    materials provided with the distribution.
 *
 * 3. The name "JDBM" must not be used to endorse or promote
 *    products derived from this Software without prior written
 *    permission of Cees de Groot.  For written permission,
 *    please contact cg@cdegroot.com.
 *
 * 4. Products derived from this Software may not be called "JDBM"
 *    nor may "JDBM" appear in their names without prior written
 *    permission of Cees de Groot.
 *
 * 5. Due credit should be given to the JDBM Project
 *    (http://jdbm.sourceforge.net/).
 *
 * THIS SOFTWARE IS PROVIDED BY THE JDBM PROJECT AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESSED OR IMPLIED WARRANTIES, INCLUDING, BUT
 * NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
 * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL
 * CEES DE GROOT OR ANY CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * Copyright 2000 (C) Cees de Groot. All Rights Reserved.
 * Copyright 2000-2001 (C) Alex Boisvert. All Rights Reserved.
 * Contributions are Copyright (C) 2000 by their associated contributors.
 *
 * $Id: RecordManager.java,v 1.5 2005/11/08 20:58:28 thompsonbry Exp $
 */

package jdbm;

import java.io.IOException;

import jdbm.helper.ISerializationHandler;
import jdbm.helper.Serializer;

/**
 *  An interface to manages records, which are uninterpreted blobs of data.
 *  <p>
 *  The set of record operations is simple: fetch, insert, update and delete.
 *  Each record is identified using a "rowid" and contains a byte[] data block.
 *  Rowids are returned on inserts and you can store them someplace safe
 *  to be able to get  back to them.  Data blocks can be as long as you wish,
 *  and may have lengths different from the original when updating.
 *
 * @author <a href="mailto:boisvert@intalio.com">Alex Boisvert</a>
 * @author <a href="cg@cdegroot.com">Cees de Groot</a>
 * @version $Id: RecordManager.java,v 1.5 2005/11/08 20:58:28 thompsonbry Exp $
 */
public interface RecordManager
{

    /**
     * Reserved slot for name directory.
     */
    public static final int NAME_DIRECTORY_ROOT = 0;


    /**
     * Recid indicating no record (e.g. null)
     */
    public static final long NULL_RECID = 0;
    
    
    /**
     * Get the underlying Record Manager or <i>this</i> if this is the
     * base record manager.
     *
     * @return underlying RecordManager or <code>null</code> if record
     * manager has been closed. 
     */
    
    public RecordManager getRecordManager();

    /**
     * Return the base underlying Record Manager or <i>this</i> if this
     * is the base record manager.
     *
     * @return base RecordManager or <code>null</code> if record
     * manager has been closed.
     */
    
    public RecordManager getBaseRecordManager();
    
    
    /**
     * Return the configured serialization handler.
     * 
     * @return The configured serialization handler and never <code>null</code>.
     * 
     * @see RecordManagerOptions#SERIALIZER
     */
    
    public ISerializationHandler getSerializationHandler();
    
    
    /**
     *  Inserts a new record using standard java object serialization.
     *
     *  @param obj the object for the new record.
     *  @return the rowid for the new record.
     *  @throws IOException when one of the underlying I/O operations fails.
     */
    public abstract long insert( Object obj )
        throws IOException;

    
    /**
     *  Inserts a new record using a custom serializer.
     *
     *  @param obj the object for the new record.
     *  @param serializer a custom serializer or <code>null</code> to use the
     *  configured serialization handler.
     *  @return the rowid for the new record.
     *  @throws IOException when one of the underlying I/O operations fails.
     */
    public abstract long insert( Object obj, Serializer serializer )
        throws IOException;


    /**
     *  Deletes a record.
     *
     *  @param recid the rowid for the record that should be deleted.
     *  @throws IOException when one of the underlying I/O operations fails.
     */
    public abstract void delete( long recid )
        throws IOException;


    /**
     *  Updates a record using standard java object serialization.
     *
     *  @param recid the recid for the record that is to be updated.
     *  @param obj the new object for the record.
     *  @throws IOException when one of the underlying I/O operations fails.
     */
    public abstract void update( long recid, Object obj )
        throws IOException;


    /**
     *  Updates a record using a custom serializer.
     *
     *  @param recid the recid for the record that is to be updated.
     *  @param obj the new object for the record.
     *  @param serializer a custom serializer or <code>null</code> to use the
     *  configured serialization handler.
     *  @throws IOException when one of the underlying I/O operations fails.
     */
    public abstract void update( long recid, Object obj, Serializer serializer )
        throws IOException;

    
    /**
     *  Fetches a record using standard java object serialization.
     *
     *  @param recid the recid for the record that must be fetched.
     *  @return the object contained in the record.
     *  @throws IOException when one of the underlying I/O operations fails.
     */
    public abstract Object fetch( long recid )
        throws IOException;


    /**
     *  Fetches a record using a custom serializer.
     *
     *  @param recid the recid for the record that must be fetched.
     *  @param serializer a custom serializer or <code>null</code> to use the
     *  configured serialization handler.
     *  @return the object contained in the record.
     *  @throws IOException when one of the underlying I/O operations fails.
     */
    public abstract Object fetch( long recid, Serializer serializer )
        throws IOException;


    /**
     *  Closes the record manager.
     *
     *  @throws IOException when one of the underlying I/O operations fails.
     */
    public abstract void close()
        throws IOException;


    /**
     *  Returns the number of slots available for "root" rowids. These slots
     *  can be used to store special rowids, like rowids that point to
     *  other rowids. Root rowids are useful for bootstrapping access to
     *  a set of data.
     */
    public abstract int getRootCount();


    /**
     *  Returns the indicated root rowid.
     *
     *  @see #getRootCount
     */
    public abstract long getRoot( int id )
        throws IOException;


    /**
     *  Sets the indicated root rowid.
     *
     *  @see #getRootCount
     */
    public abstract void setRoot( int id, long rowid )
        throws IOException;


    /**
     * Commit (make persistent) all changes since beginning of transaction.
     */
    public abstract void commit()
        throws IOException;


    /**
     * Rollback (cancel) all changes since beginning of transaction.
     */
    public abstract void rollback()
        throws IOException;




    /**
     * Obtain the record id of a named object. Returns 0 if named object
     * doesn't exist.
     */
    public abstract long getNamedObject( String name )
        throws IOException;


    /**
     * Set the record id of a named object.
     * 
     * @param The name that may be used to recover the <i>recid</i>.
     * 
     * @param The record id or <code>0</code> to remove the name from
     * the dictionary.
     */
    public abstract void setNamedObject( String name, long recid )
        throws IOException;

}
